File: Kalorie/Kalorie/Core/UseCases/AssignFoodMealTypeUseCase.swift
//
//  AssignFoodMealTypeUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 03.09.2026.
//

import Foundation

protocol AssignFoodMealTypeUseCaseProtocol {
    func callAsFunction(_ food: FoodConsumedDomain, mealTypeId: String) async throws
}

struct AssignFoodMealTypeUseCase: AssignFoodMealTypeUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(_ food: FoodConsumedDomain, mealTypeId: String) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let dto = FoodConsumedDTO(food: food, mealTypeId: mealTypeId)
        try await dataProvider.setAsync(dto, id: food.id, in: Constants.Firestore.foodConsumed(userId: userId))
    }
}

#if DEBUG
struct AssignFoodMealTypeUseCaseFake: AssignFoodMealTypeUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(_ food: FoodConsumedDomain, mealTypeId: String) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/AssignFoodMealTypeUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodConsumedDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodConsumedDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants

interface AssignFoodMealTypeUseCaseProtocol {
    suspend operator fun invoke(food: FoodConsumedDomain, mealTypeId: String)
}

class AssignFoodMealTypeUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : AssignFoodMealTypeUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(food: FoodConsumedDomain, mealTypeId: String) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val dto = FoodConsumedDTO(food = food, mealTypeId = mealTypeId)
        dataProvider.setAsync(dto, id = food.id, inCollection = Constants.Firestore.foodConsumed(userId))
    }
}
